8267155: runtime/os/TestTracePageSizes times out

tstuefe opened the pull request:
This Linux-specific test parses `/proc/self/smaps` using a dotall regular expression. If part of the expression don't match it explodes in complexity, leading to timeouts.

In our case, `VmFlags` tag was missing from smaps, which was introduced with kernel 3.8. I am actually not able to determine how slow they were; on one machine they ran for two hours before getting killed.

I tried to fiddle with the regular expression and gave up, instead opting to rewrite the parser to get a simple one-pass scan. This is way faster than before - on our old-kernel machines the tests complete in 2 minutes. On new kernels the test is a bit faster too.

In addition to rewriting the parser, I added code which copies the smaps file into the test directory before parsing it. I do this to minimize problems should the underlying proc fs content change while parsing, and to have a way to retain the parsed smaps files.

I also added a way to feed an external smaps file into the test. Of course the test would fail, but it was a way to test the parser.

Unfortunately, this does not make the test succeed. The timeouts are gone, but we have still have no way to know if TPHs are enabled or not. That is a separate issue though.

Thanks, Thomas

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267155](https://bugs.openjdk.java.net/browse/JDK-8267155): runtime/os/TestTracePageSizes times out


### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4064/head:pull/4064` \
`$ git checkout pull/4064`

Update a local copy of the PR: \
`$ git checkout pull/4064` \
`$ git pull https://git.openjdk.java.net/jdk pull/4064/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4064`

View PR using the GUI difftool: \
`$ git pr show -t 4064`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4064.diff">https://git.openjdk.java.net/jdk/pull/4064.diff</a>

</details>

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+    // (To minimize chances of concurrent modification when parsing, as well as helping with error analysis)
+    private static Path copySmaps() throws Exception {
+        Path p1 = Paths.get("/proc/self/smaps");
+        Path p2 = Paths.get("smaps-copy-" +  ProcessHandle.current().pid() + "-" + run++ + ".txt");
```
```suggestion
        Path p2 = Paths.get("smaps-copy-" +  ProcessHandle.current().pid() + "-" + (run++) + ".txt");
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+        Path p1 = Paths.get("/proc/self/smaps");
+        Path p2 = Paths.get("smaps-copy-" +  ProcessHandle.current().pid() + "-" + run++ + ".txt");
+        Files.copy(p1, p2, StandardCopyOption.REPLACE_EXISTING);
+        System.out.println("Copied " + p1.getFileName() + " to " + p2.getFileName() + "...");
```
`Path.toString()` is defined, I think, so...

```suggestion
        System.out.println("Copied " + p1 + " to " + p2 + "...");
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+        String end;
+        String ps;
+        String vmFlags;
+        Pattern sectionStartPat = Pattern.compile("^([a-f0-9]+)-([a-f0-9]+) [\\-rwpsx]{4}.*");
+        Pattern kernelPageSizePat = Pattern.compile("^KernelPageSize:\\s*(\\d*) kB");
+        Pattern vmFlagsPat = Pattern.compile("^VmFlags: ([\\w\\? ]*)");
```
```suggestion
        static final Pattern SECTION_START_PATT = Pattern.compile("^([a-f0-9]+)-([a-f0-9]+) [\\-rwpsx]{4}.*");
        static final Pattern KERNEL_PAGESIZE_PATT = Pattern.compile("^KernelPageSize:\\s*(\\d*) kB");
        static final Pattern VMFLAGS_PATT = Pattern.compile("^VmFlags: ([\\w\\? ]*)");
```

shipilev replied:
...and probably move it before instance fields block.

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+        Pattern vmFlagsPat = Pattern.compile("^VmFlags: ([\\w\\? ]*)");
+        int lineno = 0;
+        void reset() {
+            start = end = ps = vmFlags = null;
```
```suggestion
            start = null;
            end = null;
            ps = null;
            vmFlags = null;
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+        }
+
+        void eatNext(String line) {
+            debug("" + lineno++ + " " + line);
```
```suggestion
            debug("" + (lineno++) + " " + line);
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+                finish();
+                start = matSectionStart.group(1);
+                end = matSectionStart.group(2);
+                ps = vmFlags = null;
```
```suggestion
                ps = null;
                vmFlags = null;
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,87 @@
+    // Parse /proc/self/smaps
+    private static void parseSmaps(Path smapsFileToParse) throws Exception {
+        System.out.println("Parsing: " + smapsFileToParse.getFileName() + "...");
+        SmapsParser parser2 = new SmapsParser();
+        Files.lines(smapsFileToParse).forEach(parser2::eatNext);
+        parser2.finish();
```
```suggestion
        SmapsParser parser = new SmapsParser();
        Files.lines(smapsFileToParse).forEach(parser::eatNext);
        parser.finish();
```

shipilev reviewed (commented):
Cursory review follows. 

The alternative would be to disable this on kernels < 3.8? See the commits from my last PR:
https://github.com/openjdk/jdk/pull/3415/files/c7a6136cedb827260f43ecc850a8910492cfefcc..1df37de73ed6615db565aeafe76edf30dcee108d

tstuefe commented:
> Cursory review follows.
> 
> The alternative would be to disable this on kernels < 3.8? See the commits from my last PR:
> https://github.com/openjdk/jdk/pull/3415/files/c7a6136cedb827260f43ecc850a8910492cfefcc..1df37de73ed6615db565aeafe76edf30dcee108d

Hmm, I like faster tests even on new kernels, therefore I would like to use this parser instead of the regex. Also think it makes the test more readable and robust, but that may be just me. E.g. you don't rely on order of entry in smaps. 

But we could do both, of course. Lets ask  @kstefanj , its his test.

tstuefe commented:
I also reduced the heap size of the tests to 128m, which makes it more likely for the explicit huge page allocations to actually get large pages, especially when running in parallel with other tests.

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,91 @@
+    static class SmapsParser {
+        static final Pattern SECTION_START_PATT = Pattern.compile("^([a-f0-9]+)-([a-f0-9]+) [\\-rwpsx]{4}.*");
+        static final Pattern KERNEL_PAGESIZE_PATT = Pattern.compile("^KernelPageSize:\\s*(\\d*) kB");
+        static final Pattern VMFLAGS_PATT = Pattern.compile("^VmFlags: ([\\w\\? ]*)");        String start;
```
```suggestion
        static final Pattern VMFLAGS_PATT = Pattern.compile("^VmFlags: ([\\w\\? ]*)");
        String start;
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,91 @@
+        String end;
+        String ps;
+        String vmFlags;
+        int lineno = 0;
```
```suggestion
        int lineno;
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -157,8 +220,6 @@ public static void main(String args[]) throws Exception {
         // Check if debug printing is enabled.
         if (args.length > 0 && args[0].equals("-debug")) {
             debug = true;
-        } else {
-            debug = false;
```
Let's keep the code shape with explicit initialization here? Also drop the init at the field declaration?

tstuefe replied:
I did this to quickly turn on debug by changing the default value above. I could not figure out how to add command line parameters to the test on the jtreg command line and did not feel like modifying every single at-run. But I have no strong emotions, if you prefer I can reinstate the old behaviour.

shipilev replied:
Yeah, let's keep it the old way. You can change the `debug = true` right here next time, I think. IIRC, that's what I did before.

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,91 @@
     // Store address ranges with known page size.
     private static LinkedList<RangeWithPageSize> ranges = new LinkedList<>();
-    private static boolean debug;
+    private static boolean debug = false;
+    private static int run = 0;
```
```suggestion
    private static boolean debug;
    private static int run;
```

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,91 @@
+                finish();
+                start = matSectionStart.group(1);
+                end = matSectionStart.group(2);
+                ps = null;
+                vmFlags = null;
```
I can see that `Range` code handles `vmFlags = null` specifically. Does it handle `ps == null`? I cannot see it.

shipilev reviewed (commented):
Yes, I don't mind this going in. I was just pointing the way you could fix the test on <3.8 kernels either here on in separate PR.

kstefanj commented:
> > Cursory review follows.
> > The alternative would be to disable this on kernels < 3.8? See the commits from my last PR:
> > https://github.com/openjdk/jdk/pull/3415/files/c7a6136cedb827260f43ecc850a8910492cfefcc..1df37de73ed6615db565aeafe76edf30dcee108d
> 
> Hmm, I like faster tests even on new kernels, therefore I would like to use this parser instead of the regex. Also think it makes the test more readable and robust, but that may be just me. E.g. you don't rely on order of entry in smaps.
> 
> But we could do both, of course. Lets ask @kstefanj , its his test.

I'm of course biased since I wrote the test. That said, I don't consider this _my_ test, so I won't object to changing it. 

Personally I prefer the regex-parsing is clearer since it "parses" one segment at a time in the smaps file. I think if we go with this new parser, it would be nice to add some comments clarifying the steps. 

An alternative would be to have different regular expressions depending on the kernel version or first check if vmFlags are present and choose the correct expression. This would of course not help if this kind of parsing is still very slow on some systems.

I like that you do a copy of smaps before parsing, that crossed my mind as well, to help debugging. Regarding smaller heap sizes. I don't see a problem if some of the test VMs fail getting large pages, since it tests that we still log the correct thing. 

Just wanted to share my thoughts before doing a proper review. I'm happy to review this patch if we decide to go with this new parser. If we do that I also think we should look at enabling parsing of `AnonHugePages` for the case where vmFlags are not present (similar to @shipilev approach in the last PR).

tstuefe commented:
Hi Stefan,

> > > Cursory review follows.
> > > The alternative would be to disable this on kernels < 3.8? See the commits from my last PR:
> > > https://github.com/openjdk/jdk/pull/3415/files/c7a6136cedb827260f43ecc850a8910492cfefcc..1df37de73ed6615db565aeafe76edf30dcee108d
> > 
> > 
> > Hmm, I like faster tests even on new kernels, therefore I would like to use this parser instead of the regex. Also think it makes the test more readable and robust, but that may be just me. E.g. you don't rely on order of entry in smaps.
> > But we could do both, of course. Lets ask @kstefanj , its his test.
> 
> I'm of course biased since I wrote the test. That said, I don't consider this _my_ test, so I won't object to changing it.
> 

Sorry, I meant no offence. I think this test is really useful, since I want to be able to rely on the page sizes in the traces and not have to countercheck the smaps file. That's why I put work in it instead of just disabling it.

What I meant was do you consider testing correct tracing of *explicit* huge pages important enough on semi-old kernels before 3.8? If not, we can modify the test to skip older kernels completely, otherwise just the detect-THP-part?

> Personally I prefer the regex-parsing is clearer since it "parses" one segment at a time in the smaps file.

Unfortunately its not so easy since its smaps sections don't have a clear ending. I played around yesterday and gave up after a while. I also find regex code beyond a certain complexity really work intensive and difficult to debug.

The original regexp was also too loose, I think, since `(\\w+)-(\\w+)` - unanchored to the start of a line - matches the permissions if those contain dashes. And theoretically, you could accidentally match across secions, were one to miss a flag.

> I think if we go with this new parser, it would be nice to add some comments clarifying the steps.

Okay, I could do that.

> 
> An alternative would be to have different regular expressions depending on the kernel version or first check if vmFlags are present and choose the correct expression. This would of course not help if this kind of parsing is still very slow on some systems.

I think its just more vulnerable against unknown or future smaps variants. E.g. relying on a specific order of entries.

> 
> I like that you do a copy of smaps before parsing, that crossed my mind as well, to help debugging. Regarding smaller heap sizes. I don't see a problem if some of the test VMs fail getting large pages, since it tests that we still log the correct thing.

I prefer the "explicit-large-pages" test to really run a VM with explicit large pages. Otherwise we test with small pages, which would be the same as with the other tests.

> 
> Just wanted to share my thoughts before doing a proper review. I'm happy to review this patch if we decide to go with this new parser. If we do that I also think we should look at enabling parsing of `AnonHugePages` for the case where vmFlags are not present (similar to @shipilev approach in the last PR).

This test has been a bit of whack-a-mole; I am interested in a stable test which covers the important cases. E.g. I tried to find a clear definition of AnonHugePages but beyond that it requires the kernel to be compiled with THP I did not find any indication that this is synonymous with THP, or under which circumstances it is set, or whats its relationship to KernelPageSize is. I also suspect it is kernel-version-dependent like the other properties.

Frankly am interested in this test for testing trace of explicit huge pages correctly, not so much THP. Since this is what most of our customers use. THP have a bad rep here because they used to come with problems. And since hugetlbfs has gotten easier to use (eg with vm.nr_overcommit_hugepages) I usually use that instead of TPH.

Cheers, Thomas

shipilev commented:
> > Just wanted to share my thoughts before doing a proper review. I'm happy to review this patch if we decide to go with this new parser. If we do that I also think we should look at enabling parsing of `AnonHugePages` for the case where vmFlags are not present (similar to @shipilev approach in the last PR).
> 
> This test has been a bit of whack-a-mole; I am interested in a stable test which covers the important cases. E.g. I tried to find a clear definition of AnonHugePages but beyond that it requires the kernel to be compiled with THP I did not find any indication that this is synonymous with THP, or under which circumstances it is set, or whats its relationship to KernelPageSize is.

https://www.kernel.org/doc/Documentation/vm/transhuge.txt:
"The number of anonymous transparent huge pages currently used by the system is available by reading the AnonHugePages field in /proc/meminfo. To identify what applications are using anonymous transparent huge pages, it is necessary to read /proc/PID/smaps and count the AnonHugePages fields for each mapping."

But as we found earlier, this does not work because for many VM structures we are only _reserving_ the memory, while AnonHugePages > 0 only when we actually _commit_ stuff. So AnonHugePages is a lost cause, IMO. The only reliable way to test THP seems to be VmFlags. If VmFlags are not available, that the end. We can only stop testing for THP at that point...

kstefanj commented:
> > 
> > I'm of course biased since I wrote the test. That said, I don't consider this _my_ test, so I won't object to changing it.
> 
> Sorry, I meant no offence. I think this test is really useful, since I want to be able to rely on the page sizes in the traces and not have to countercheck the smaps file. That's why I put work in it instead of just disabling it.
> 
None taken 😄 

> What I meant was do you consider testing correct tracing of _explicit_ huge pages important enough on semi-old kernels before 3.8? If not, we can modify the test to skip older kernels completely, otherwise just the detect-THP-part?
> 
I'm good with keeping testing even on older kernels, since you run into this it obviously has some value testing it.

> > Personally I prefer the regex-parsing is clearer since it "parses" one segment at a time in the smaps file.
> 
> Unfortunately its not so easy since its smaps sections don't have a clear ending. I played around yesterday and gave up after a while. I also find regex code beyond a certain complexity really work intensive and difficult to debug.
> 
> The original regexp was also too loose, I think, since `(\\w+)-(\\w+)` - unanchored to the start of a line - matches the permissions if those contain dashes. And theoretically, you could accidentally match across secions, were one to miss a flag.
> 
True, the reg-ex approach has problems and as you say, future changes to the format might be more brittle with a reg-ex. 

> > I think if we go with this new parser, it would be nice to add some comments clarifying the steps.
> 
> Okay, I could do that.
> 
👍 

> > An alternative would be to have different regular expressions depending on the kernel version or first check if vmFlags are present and choose the correct expression. This would of course not help if this kind of parsing is still very slow on some systems.
> 
> I think its just more vulnerable against unknown or future smaps variants. E.g. relying on a specific order of entries.
> 
> > I like that you do a copy of smaps before parsing, that crossed my mind as well, to help debugging. Regarding smaller heap sizes. I don't see a problem if some of the test VMs fail getting large pages, since it tests that we still log the correct thing.
> 
> I prefer the "explicit-large-pages" test to really run a VM with explicit large pages. Otherwise we test with small pages, which would be the same as with the other tests.
> 
What I meant was that we can get the scenario where some reservations are small and some large, but that can still happen even with a smaller heap size, so I'm good with changing it.

> > Just wanted to share my thoughts before doing a proper review. I'm happy to review this patch if we decide to go with this new parser. If we do that I also think we should look at enabling parsing of `AnonHugePages` for the case where vmFlags are not present (similar to @shipilev approach in the last PR).
> 
> This test has been a bit of whack-a-mole; I am interested in a stable test which covers the important cases. E.g. I tried to find a clear definition of AnonHugePages but beyond that it requires the kernel to be compiled with THP I did not find any indication that this is synonymous with THP, or under which circumstances it is set, or whats its relationship to KernelPageSize is. I also suspect it is kernel-version-dependent like the other properties.
> 
> Frankly am interested in this test for testing trace of explicit huge pages correctly, not so much THP. Since this is what most of our customers use. THP have a bad rep here because they used to come with problems. And since hugetlbfs has gotten easier to use (eg with vm.nr_overcommit_hugepages) I usually use that instead of TPH.
> 

Yeah, the THP part might have been over-shooting it a bit, but I was quite happy when I found the `vmFlags` and never considered the older kernels. I'm good with disabling THP testing when vmFlags are not present.

kstefanj commented:
> But as we found earlier, this does not work because for many VM structures we are only _reserving_ the memory, while AnonHugePages > 0 only when we actually _commit_ stuff. So AnonHugePages is a lost cause, IMO. The only reliable way to test THP seems to be VmFlags. If VmFlags are not available, that the end. We can only stop testing for THP at that point...

Correct, forgot about that =) Let's just skip testing of THP if vmFlags are missing.

kstefanj commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,92 @@
+        Path p1 = Paths.get("/proc/self/smaps");
+        Path p2 = Paths.get("smaps-copy-" +  ProcessHandle.current().pid() + "-" + (run++) + ".txt");
+        Files.copy(p1, p2, StandardCopyOption.REPLACE_EXISTING);
+        System.out.println("Copied " + p1 + " to " + p2 + "...");
```
Consider using `debug(...)` for this and the other `System.out.println(..)`calls.

kstefanj commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -93,32 +97,92 @@
-            debug("Added range: " + range);
-        });
-        smapsScanner.close();
+        // We can override the smaps file to parse to pass in a pre-fetched one
```
Please update the above comment to match the new parsing.

kstefanj commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -246,11 +308,13 @@ public RangeWithPageSize(String start, String end, String pageSize, String vmFla
-                vmFlagHT = true;
-            } else if (flag.equals("hg")) {
-                vmFlagHG = true;
+        if (vmFlags != null) {
```
Update comment to explain the smaps sometimes don't have `vmFlags`.

kstefanj reviewed (changes requested):
Added some initial comments to the code.

tstuefe commented:
The more I think about this the more confused I get.

In theory, THP can be enabled on a machine in a way which needs no cooperation of the VM; so whether or not we run with +UseTransparentHugePages makes no difference, pages used by the VM may be folded transparently to huge pages without our knowledge. Would on those machines the section show up as "VmFlags: hg" too? Or is this only the explicit madvice flag? Would the KernelPageSize reflect the page size set by the THP folding process (probably not since it can be multiple page sizes)?

And would this "proactive folding" interfere with the recognition of hugetlb pages? In other words, could we trace out "4k" and then have a "2048k" region show up in smaps?

The more I think about this, the muddier it gets. ATM I think we should just skip this test altogether for older kernels, and/or not test THP at all... not sure what to think here.

I also found this piece in man proc:
```
The "KernelPageSize" line (available since Linux 2.6.29)
 is the page size used by the kernel to back the virtual
 memory area.  .....  
However, one counter-example occurs on PPC64 kernels 
whereby a kernel using 64 kB as a base page
 size may still use 4 kB pages for the MMU on older
 processors.
```
I propose to just skip the test on that platform. Personally I have never used large pages on a ppc machine.

kstefanj commented:
> The more I think about this the more confused I get.
> 
I recognize that feeling 😄 

> In theory, THP can be enabled on a machine in a way which needs no cooperation of the VM; so whether or not we run with +UseTransparentHugePages makes no difference, pages used by the VM may be folded transparently to huge pages without our knowledge. Would on those machines the section show up as "VmFlags: hg" too? Or is this only the explicit madvice flag? Would the KernelPageSize reflect the page size set by the THP folding process (probably not since it can be multiple page sizes)?
> 
From what I've seen KernelPageSize is only different for the HugeTLBFS pages. Reading the doc it says this about the `hg`flag:
`hg  - huge page advise flag`

So I also think it would only show when running with `+UseTransparentHugePages`.

> And would this "proactive folding" interfere with the recognition of hugetlb pages? In other words, could we trace out "4k" and then have a "2048k" region show up in smaps?
> 
> The more I think about this, the muddier it gets. ATM I think we should just skip this test altogether for older kernels, and/or not test THP at all... not sure what to think here.
> 
I'm totally fine with skipping older kernel. I think the fact that we test this on newer kernels and it passes at least shows that we trace the correct thing.

> I also found this piece in man proc:
> 
> ```
> The "KernelPageSize" line (available since Linux 2.6.29)
>  is the page size used by the kernel to back the virtual
>  memory area.  .....  
> However, one counter-example occurs on PPC64 kernels 
> whereby a kernel using 64 kB as a base page
>  size may still use 4 kB pages for the MMU on older
>  processors.
> ```
> 
> I propose to just skip the test on that platform. Personally I have never used large pages on a ppc machine.

Fine with me :)

kstefanj commented:
Did a quick test now with THP configured to `always` but running with `-XX:-UseLargePages -XX:-UseTransparentHugePages`. It gives me:
```
80000000-ffe00000 rw-p 00000000 00:00 0 
Size:            2095104 kB
KernelPageSize:        4 kB
MMUPageSize:           4 kB
...
AnonHugePages:   2095104 kB
THPeligible:            1
VmFlags: rd wr mr mw me ac sd 
```
So no `hg` flag as expected.

tstuefe commented:
> >   
> > I propose to just skip the test on that platform. Personally I have never used large pages on a ppc machine.
> 
> Fine with me :)

Great, at least we can rely on a set of sane information from smaps, and it will make the test simpler too.

Cheers, Thomas

tstuefe commented:
Hi @kstefanj , @shipilev ,

I changed the test so that:
- we skip out if kernel < 3.8. This means we can rely on a sane set of tags from the smaps file.
- I skipped also for ppc because I found that the semantics of some tags in smaps differ and I honestly never used hugetlbfs on ppc; if it becomes an issue, we can spend more time on this.
- I reverted some of my earlier changes; now we again require all information to be there (so no null handling for missing smaps information in `RangeWithPageSize`.
- added some comments

Note that even though we skip out of the test we still at least start the VMs and go through VM initialization. That is totally fine by me since this is a good test that the VM survives these variations of large page flags (eg we found unrelated crashes on ppc with https://github.com/openjdk/jdk/pull/1153 that way).

Cheers, Thomas

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -94,31 +112,97 @@
-    // transparent huge pages. The pattern is not greedy and will
-    // match as little as possible so each "segment" in the file
-    // will generate a match.
+    private static long getKernelVersion() {
+        long kernelVersion = Platform.getOsVersionMajor() << 8 | Platform.getOsVersionMinor();
+        debug("kernelVersion " + Long.toHexString(kernelVersion));
+        return kernelVersion;
+    }
```
Let's not do any bit magic? Just inline this to the only use and do the proper branchy "if"?

kstefanj replied:
+1

shipilev commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -237,10 +335,11 @@ private static void debug(String str) {
         this.start = Long.parseUnsignedLong(start, 16);
         this.end = Long.parseUnsignedLong(end, 16);
         this.pageSize = Long.parseLong(pageSize);
-
```
Restore this newline?

shipilev reviewed (changes requested):
This looks fine to me, modulo a few nits.

kstefanj reviewed (approved):
Looks good.

kstefanj commented on test/hotspot/jtreg/runtime/os/TestTracePageSizes.java:
```diff
@@ -33,59 +35,75 @@
- * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseLargePages -XX:LargePageSizeInBytes=2m TestTracePageSizes
- * @run main/othervm -XX:+AlwaysPreTouch -Xlog:pagesize:ps-%p.log -XX:+UseLargePages -XX:LargePageSizeInBytes=1g TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xmx128m -Xlog:pagesize:ps-%p.log -XX:+UseLargePages -XX:LargePageSizeInBytes=2m TestTracePageSizes
+ * @run main/othervm -XX:+AlwaysPreTouch -Xmx128m -Xlog:pagesize:ps-%p.log -XX:+UseLargePages -XX:LargePageSizeInBytes=1g TestTracePageSizes
```
Just realized that this test should probably not set `-Xmx128m` since we try to use 1g pages. What will happen is that the heap will be rounded up to 1G (the alignment logic here is a bit strange) so we will still test it, but it feels a bit backwards. Maybe set the heap to 1g in this case?

tstuefe replied:
Good point.

tstuefe commented:
Final changes. Nobody liked my bit shifting, so removed. I upped the heap size to 2g for the largepagesizeinbytes=1g page test.

One nice thing about the new parser is that with debug=true we now can observe the parsing, including the content of the parsed smaps file in the jtr output.

shipilev reviewed (approved):
Looks fine to me.

kstefanj reviewed (approved):
Looks good!

tstuefe commented:
Thanks Alexey and Stefan.

/integrate
